File: web/src/lib/invoice/core/executor.ts
/**
 * Turn an intent + selections into a computed draft — ported from
 * src/sautice/invoice/executor.py. Every figure is integer kobo computed here,
 * never guessed; a confusable/missing entity or ambiguous amount becomes a
 * question. Output shape matches web/src/lib/invoice/bridge.ts `Draft`.
 */
import type { Draft, DraftLine, DraftQuestion } from "../bridge";
import { formatKobo, koboToNairaString, mulKobo, MoneyError, nairaToKobo } from "./money";
import { readAmount, readQuantity, readTermDays, isAmbiguous, single } from "./naira";
import {
  findCustomer, findProduct, resolveCustomer, resolveProduct,
  type Customer, type Product, type Roster,
} from "./roster";
import type { Intent } from "./extract";

export interface Selections {
  customer_id?: string;
  customer_new_name?: string;
  lines?: Array<{ product_id?: string; product_new_name?: string; qty?: number; unit_price_naira?: string }>;
}

export function buildDraft(intent: Intent, roster: Roster, selections: Selections = {}, today?: string): Draft {
  const customers = roster.customers ?? [];
  const products = roster.products ?? [];
  const questions: DraftQuestion[] = [];

  const customer = resolveCustomerField(intent, selections, customers, questions);
  const lineSel = selections.lines ?? [];
  const lines: DraftLine[] = intent.lines.map((ln, i) => buildLine(i, ln, lineSel[i] ?? {}, products, questions));

  let totalKobo = 0;
  let priced = true;
  for (const ln of lines) {
    if (ln.line_total_kobo == null) priced = false;
    else totalKobo += ln.line_total_kobo;
  }

  const terms = resolveTerms(intent.terms_text, today);
  const ready = questions.length === 0 && lines.length > 0 && priced && customer.status === "resolved";

  return {
    customer,
    lines,
    terms,
    total_kobo: priced && lines.length ? totalKobo : null,
    total_display: priced && lines.length ? formatKobo(totalKobo) : null,
    questions,
    ready,
  };
}

function resolveCustomerField(intent: Intent, selections: Selections, customers: Customer[], questions: DraftQuestion[]): Draft["customer"] {
  const cid = selections.customer_id;
  if (cid === "NEW") {
    const name = (selections.customer_new_name || intent.customer_query || "").trim();
    return { status: "new", query: intent.customer_query, resolved: { id: null, name, new: true } };
  }
  if (cid) {
    const found = findCustomer(cid, customers);
    if (found) return { status: "resolved", query: intent.customer_query, resolved: found };
  }

  const res = resolveCustomer(intent.customer_query, customers);
  if (res.status === "ambiguous") {
    questions.push({
      id: "customer", kind: "choice", field: "customer",
      prompt: `Which customer did you mean by '${res.query}'?`,
      options: [...res.candidates.map((c) => ({ value: c.id, label: c.name })), { value: "NEW", label: `New customer: ${res.query}` }],
    });
  } else if (res.status === "unknown" || res.status === "missing") {
    const q = res.query || "";
    questions.push({
      id: "customer", kind: q ? "confirm_new" : "text", field: "customer",
      prompt: q ? `'${q}' isn't in your customers. Add as a new customer?` : "Who is this invoice for?",
      options: q ? [{ value: "NEW", label: `New customer: ${q}` }] : [],
    });
  }
  return { status: res.status, query: res.query, resolved: res.resolved, candidates: res.candidates };
}

function buildLine(i: number, line: Intent["lines"][number], sel: NonNullable<Selections["lines"]>[number], products: Product[], questions: DraftQuestion[]): DraftLine {
  // --- product ---
  let product: DraftLine["product"];
  const pid = sel.product_id;
  if (pid === "NEW") {
    product = { status: "new", resolved: { id: null, name: sel.product_new_name || line.product_query, new: true, unit_price_kobo: null } };
  } else if (pid) {
    const found = findProduct(pid, products);
    product = found ? { status: "resolved", resolved: found } : resolveProduct(line.product_query, products);
  } else {
    product = resolveProduct(line.product_query, products);
  }

  if (product.status === "ambiguous") {
    questions.push({
      id: `line:${i}:product`, kind: "choice", field: "product", line: i,
      prompt: `Which product did you mean by '${line.product_query}'?`,
      options: ((product.candidates ?? []) as Product[]).map((p) => ({ value: p.id, label: `${p.name} - ${formatKobo(p.unit_price_kobo)}` })),
    });
  } else if (product.status === "unknown" || product.status === "missing") {
    questions.push({
      id: `line:${i}:product`, kind: line.product_query ? "confirm_new" : "text", field: "product", line: i,
      prompt: line.product_query ? `'${line.product_query}' isn't in your catalogue. Add it?` : "What product is this line?",
      options: line.product_query ? [{ value: "NEW", label: `New product: ${line.product_query}` }] : [],
    });
  }
  const resolvedProduct = product.resolved ?? null;

  // --- quantity ---
  let qty: number | null = null;
  if (sel.qty != null) qty = Math.floor(Number(sel.qty));
  else {
    const r = line.qty_text ? readQuantity(line.qty_text) : null;
    if (!r || r.candidates.length === 0) {
      questions.push({ id: `line:${i}:qty`, kind: "number", field: "qty", line: i, prompt: `How many for '${line.product_query || "this item"}'?` });
    } else if (isAmbiguous(r)) {
      questions.push({ id: `line:${i}:qty`, kind: "choice", field: "qty", line: i, prompt: `What quantity - ${r.note}?`, options: r.candidates.map((c) => ({ value: c, label: String(c) })) });
    } else qty = single(r);
  }

  // --- unit price ---
  let unitPriceKobo: number | null = null;
  let unitPriceSource: DraftLine["unit_price_source"] = null;
  if (sel.unit_price_naira != null) {
    try { unitPriceKobo = nairaToKobo(String(sel.unit_price_naira)); unitPriceSource = "corrected"; }
    catch (e) { if (e instanceof MoneyError) questions.push({ id: `line:${i}:price`, kind: "amount", field: "price", line: i, prompt: "That price didn't read as a naira amount. What is the unit price?" }); else throw e; }
  } else if (line.price_text) {
    const r = readAmount(line.price_text);
    if (r.candidates.length === 0) {
      questions.push({ id: `line:${i}:price`, kind: "amount", field: "price", line: i, prompt: `What is the unit price for '${line.product_query || "this item"}'?` });
    } else if (isAmbiguous(r)) {
      questions.push({ id: `line:${i}:price`, kind: "choice", field: "price", line: i, prompt: `What price - ${r.note}?`, options: r.candidates.map((c) => ({ value: koboToNairaString(c), label: formatKobo(c) })) });
    } else { unitPriceKobo = single(r); unitPriceSource = "spoken"; }
  } else if (resolvedProduct && resolvedProduct.unit_price_kobo != null) {
    unitPriceKobo = resolvedProduct.unit_price_kobo;
    unitPriceSource = "catalog";
  } else if (product.status !== "ambiguous" && product.status !== "unknown" && product.status !== "missing") {
    questions.push({ id: `line:${i}:price`, kind: "amount", field: "price", line: i, prompt: `What is the unit price for '${resolvedProduct?.name || line.product_query || "this item"}'?` });
  }

  const lineTotalKobo = qty != null && unitPriceKobo != null ? mulKobo(unitPriceKobo, qty) : null;

  return {
    index: i,
    query: line.product_query,
    product,
    qty,
    unit_price_kobo: unitPriceKobo,
    unit_price_display: unitPriceKobo != null ? formatKobo(unitPriceKobo) : null,
    unit_price_source: unitPriceSource,
    line_total_kobo: lineTotalKobo,
    line_total_display: lineTotalKobo != null ? formatKobo(lineTotalKobo) : null,
  };
}

function resolveTerms(termsText: string, today?: string): Draft["terms"] {
  if (!termsText) return { text: "", days: null, due_date: null };
  const days = readTermDays(termsText);
  const base = parseToday(today);
  let due: string | null = null;
  if (typeof days === "number") {
    const d = new Date(base.getTime() + days * 86_400_000);
    due = d.toISOString().slice(0, 10);
  } else if (days === "end_of_month") {
    const last = new Date(Date.UTC(base.getUTCFullYear(), base.getUTCMonth() + 1, 0));
    due = last.toISOString().slice(0, 10);
  }
  return { text: termsText, days: typeof days === "number" ? days : null, rule: typeof days === "string" ? days : null, due_date: due };
}

function parseToday(today?: string): Date {
  if (today && /^\d{4}-\d{2}-\d{2}$/.test(today)) return new Date(today + "T00:00:00Z");
  return new Date();
}

File: src/sautice/invoice/executor.py
"""Turn an intent + the user's selections into a fully computed draft.

Rules that matter:
  - Every naira figure is an integer-kobo Money value computed here, never in the
    model or the browser.
  - A confusable or missing entity, or an ambiguous amount, becomes a question.
    The draft is `ready` only when nothing is left to ask.
  - A price spoken in the sentence overrides the catalogue price, and is labelled
    as such so the reviewer can see it.
"""
from __future__ import annotations

import calendar
from datetime import date, timedelta
from typing import Any

from sautice.core.money import Money, MoneyError
from sautice.nlp.naira import read_amount, read_quantity, read_term_days

from .extract import Intent
from .roster import find_customer, find_product, resolve_customer, resolve_product


def build_draft(intent: Intent, roster: dict, selections: dict | None = None,
                today: str | None = None) -> dict[str, Any]:
    selections = selections or {}
    customers = roster.get("customers", [])
    products = roster.get("products", [])
    questions: list[dict] = []

    customer = _resolve_customer(intent, selections, customers, questions)
    line_sel = selections.get("lines", [])
    lines: list[dict] = []
    for i, ln in enumerate(intent.lines):
        sel = line_sel[i] if i < len(line_sel) else {}
        lines.append(_build_line(i, ln, sel, products, questions))

    total_kobo = 0
    priced = True
    for ln in lines:
        if ln["line_total_kobo"] is None:
            priced = False
        else:
            total_kobo += ln["line_total_kobo"]

    terms = _resolve_terms(intent.terms_text, today)

    # A customer the user explicitly confirmed as new is as good as one from the
    # roster: the draft must be creatable, or "New customer" would be a dead end.
    ready = not questions and bool(lines) and priced and customer["status"] in ("resolved", "new")
    return {
        "customer": customer,
        "lines": lines,
        "terms": terms,
        "total_kobo": total_kobo if priced and lines else None,
        "total_display": Money.from_kobo(total_kobo).format() if priced and lines else None,
        "questions": questions,
        "ready": ready,
    }


def _resolve_customer(intent: Intent, selections: dict, customers: list, questions: list) -> dict:
    cid = selections.get("customer_id")
    if cid == "NEW":
        name = (selections.get("customer_new_name") or intent.customer_query or "").strip()
        return {"status": "new", "query": intent.customer_query, "resolved": {"id": None, "name": name, "new": True}}
    if cid:
        found = find_customer(cid, customers)
        if found:
            return {"status": "resolved", "query": intent.customer_query, "resolved": found}

    res = resolve_customer(intent.customer_query, customers)
    if res["status"] == "ambiguous":
        questions.append({
            "id": "customer", "kind": "choice", "field": "customer",
            "prompt": f"Which customer did you mean by '{res['query']}'?",
            "options": [{"value": c["id"], "label": c["name"]} for c in res["candidates"]]
                       + [{"value": "NEW", "label": f"New customer: {res['query']}"}],
        })
    elif res["status"] in ("unknown", "missing"):
        q = res.get("query") or ""
        questions.append({
            "id": "customer", "kind": "confirm_new" if q else "text", "field": "customer",
            "prompt": (f"'{q}' isn't in your customers. Add as a new customer?" if q
                       else "Who is this invoice for?"),
            "options": ([{"value": "NEW", "label": f"New customer: {q}"}] if q else []),
        })
    return res


def _build_line(i: int, line, sel: dict, products: list, questions: list) -> dict:
    # --- product ---
    pid = sel.get("product_id")
    if pid == "NEW":
        product = {"status": "new", "resolved": {"id": None, "name": sel.get("product_new_name") or line.product_query, "new": True, "unit_price_kobo": None}}
    elif pid:
        found = find_product(pid, products)
        product = {"status": "resolved", "resolved": found} if found else resolve_product(line.product_query, products)
    else:
        product = resolve_product(line.product_query, products)

    if product["status"] == "ambiguous":
        questions.append({
            "id": f"line:{i}:product", "kind": "choice", "field": "product", "line": i,
            "prompt": f"Which product did you mean by '{line.product_query}'?",
            "options": [{"value": p["id"], "label": f"{p['name']} - {Money.from_kobo(p['unit_price_kobo']).format()}"} for p in product["candidates"]],
        })
    elif product["status"] in ("unknown", "missing"):
        questions.append({
            "id": f"line:{i}:product", "kind": "confirm_new" if line.product_query else "text", "field": "product", "line": i,
            "prompt": (f"'{line.product_query}' isn't in your catalogue. Add it?" if line.product_query
                       else "What product is this line?"),
            "options": ([{"value": "NEW", "label": f"New product: {line.product_query}"}] if line.product_query else []),
        })

    resolved_product = product.get("resolved")

    # --- quantity ---
    qty = None
    if sel.get("qty") is not None:
        qty = int(sel["qty"])
    else:
        r = read_quantity(line.qty_text) if line.qty_text else None
        if r is None or not r.candidates:
            questions.append({"id": f"line:{i}:qty", "kind": "number", "field": "qty", "line": i,
                              "prompt": f"How many for '{line.product_query or 'this item'}'?"})
        elif r.is_ambiguous:
            questions.append({"id": f"line:{i}:qty", "kind": "choice", "field": "qty", "line": i,
                              "prompt": f"What quantity - {r.note}?",
                              "options": [{"value": c, "label": str(c)} for c in r.candidates]})
        else:
            qty = r.value

    # --- unit price ---
    unit_price_kobo = None
    unit_price_source = None
    if sel.get("unit_price_naira") is not None:
        try:
            unit_price_kobo = Money.from_naira(str(sel["unit_price_naira"])).kobo
            unit_price_source = "corrected"
        except MoneyError:
            questions.append({"id": f"line:{i}:price", "kind": "amount", "field": "price", "line": i,
                              "prompt": f"That price didn't read as a naira amount. What is the unit price?"})
    elif line.price_text:
        r = read_amount(line.price_text)
        if not r.candidates:
            questions.append({"id": f"line:{i}:price", "kind": "amount", "field": "price", "line": i,
                              "prompt": f"What is the unit price for '{line.product_query or 'this item'}'?"})
        elif r.is_ambiguous:
            questions.append({"id": f"line:{i}:price", "kind": "choice", "field": "price", "line": i,
                              "prompt": f"What price - {r.note}?",
                              "options": [{"value": str(c.naira), "label": c.format()} for c in r.candidates]})
        else:
            unit_price_kobo = r.value.kobo
            unit_price_source = "spoken"
    elif resolved_product and resolved_product.get("unit_price_kobo") is not None:
        unit_price_kobo = resolved_product["unit_price_kobo"]
        unit_price_source = "catalog"
    else:
        # No spoken price and no catalogue price (e.g. a new product).
        if product["status"] not in ("ambiguous", "unknown", "missing"):
            questions.append({"id": f"line:{i}:price", "kind": "amount", "field": "price", "line": i,
                              "prompt": f"What is the unit price for '{(resolved_product or {}).get('name') or line.product_query or 'this item'}'?"})

    line_total_kobo = None
    if qty is not None and unit_price_kobo is not None:
        line_total_kobo = (Money.from_kobo(unit_price_kobo) * qty).kobo

    return {
        "index": i,
        "query": line.product_query,
        "product": product,
        "qty": qty,
        "unit_price_kobo": unit_price_kobo,
        "unit_price_display": Money.from_kobo(unit_price_kobo).format() if unit_price_kobo is not None else None,
        "unit_price_source": unit_price_source,
        "line_total_kobo": line_total_kobo,
        "line_total_display": Money.from_kobo(line_total_kobo).format() if line_total_kobo is not None else None,
    }


def _resolve_terms(terms_text: str, today: str | None) -> dict:
    if not terms_text:
        return {"text": "", "days": None, "due_date": None}
    days = read_term_days(terms_text)
    base = _parse_today(today)
    due = None
    if isinstance(days, int):
        due = (base + timedelta(days=days)).isoformat()
    elif days == "end_of_month":
        last = calendar.monthrange(base.year, base.month)[1]
        due = date(base.year, base.month, last).isoformat()
    return {"text": terms_text, "days": days if isinstance(days, int) else None,
            "rule": days if isinstance(days, str) else None, "due_date": due}


def _parse_today(today: str | None) -> date:
    if today:
        try:
            return date.fromisoformat(today)
        except ValueError:
            pass
    return date.today()
